### DISCUSSION TITLE
Sponge Source term at the boundaries

### DISCUSSION THREAD

**Discussion by cjlauer16** [AMReX-Combustion/PeleC#759] [Q&A] [answered]
I want to add a source term near the inlet and outlet to dampen out any waves and keep the boundaries more stable and am having trouble finding an example for how to do that. Can you help point me in the right direction for how to implement this?

**Comment by baperry2** [comment, marked as answer]
For now, to add arbitrary source terms, the procedure is to modify the `fill_ext_source` function as needed: https://github.com/AMReX-Combustion/PeleC/blob/33ad884cbcca0642405a61eef352b9c253f01213/Source/External.cpp#L40

When #743 gets merged, custom source terms will be able to be added by defining a function in the `prob` files, and there may even be similar capability designed for sponge zones!

**Reply by cjlauer16** [reply]
Thank you! That's great for now, and I'll keep an eye out for the new capabilities.

**Reply by cjlauer16** [reply]
What is the best way to get coordinate information into the fill_ext_source function? I have a sponge source set up based on the (i,j,k) indices used to loop over the mesh, but I'd like to use (x,y,z) coordinate values instead to better deal with changing meshes when running with amr. Thank you!

**Reply by baperry2** [reply]
We typically manually compute this as needed based on the i,j,k index and the the AMReX geometry data. Use `const auto geomdata = geom.data();` to get the GPU-safe AMReX GeometryData from the Geometry object that is part of the PeleC class. Then compute the coordinate values as is done here: https://github.com/AMReX-Combustion/PeleC/blob/3cef3cae90feafb8f9d190f3d1bbaf1a8fc8ca13/Source/Derive.cpp#L757

**Reply by baperry2** [reply]
#743 has now been merged. You can now create custom source terms with the `problem_modify_ext_sources` function or add artificial viscosity using the `problem_modify_transport_coeffs` function. The `EB-ConvergingNozzle` case now has an example of using the latter approach to dampen waves at the boundaries - probably not the ideal approach for that but it did substantially reduce spurious wave reflections happening in that case.
